src/st/st-widget.c

```diff
@@ -1540,6 +1563,61 @@ on_transition_completed (StThemeNodeTransition *transition,
   st_widget_remove_transition (widget);
 }
 
+static void st_widget_add_background_effects (StWidget    *widget,
+                                              StThemeNode *old_theme_node )
+{
+StThemeNode *new_theme_node = st_widget_get_theme_node (widget);
```

Comment:
We could probably make `st_widget_recompute_style` return the `new_theme_node` passed as an argument to this function and avoid a function invocation in the paint cycle.

Reply:
Yes, but this function is also called without being preceded by recompute_style, so that won't work in that instance.

Reply:
Sounds like a lot of overhead for a feature maybe 25% of Cinnamon users will take advantage of (optimistically).

Reply:
Yes, but remember I am only re-implementing something that was already there and had to be turned off last time because it was broken.  There's no hurry as far as I am concerned.  With all the major performance / latency stuff that you've done, the last thing I want to do is to compromise that.  So right now I am waiting for the dust to settle, maybe next release, before looking at this PR again and seeing how it might be improved.